dropshot/src/handler.rs

```diff
@@ -330,26 +335,31 @@ macro_rules! impl_HttpHandlerFunc_for_func_with_params {
     ($(($i:tt, $T:tt)),*) => {
 
     #[async_trait]
-    impl<Context, FuncType, FutureType, ResponseType, $($T,)*>
+    impl<Context, FuncType, FutureType, ResponseType, ErrorType, $($T,)*>
         HttpHandlerFunc<Context, ($($T,)*), ResponseType> for FuncType
     where
         Context: ServerContext,
         FuncType: Fn(RequestContext<Context>, $($T,)*)
             -> FutureType + Send + Sync + 'static,
-        FutureType: Future<Output = Result<ResponseType, HttpError>>
+        FutureType: Future<Output = Result<ResponseType, ErrorType>>
             + Send + 'static,
         ResponseType: HttpResponse + Send + Sync + 'static,
+        ErrorType: IntoErrorResponse
+            + std::fmt::Display
+            + Send
+            + 'static,
         ($($T,)*): RequestExtractor,
         $($T: Send + Sync + 'static,)*
     {
+        type Error = ErrorType;
         async fn handle_request(
             &self,
             rqctx: RequestContext<Context>,
             _param_tuple: ($($T,)*)
         ) -> HttpHandlerResult
         {
             let response: ResponseType =
-                (self)(rqctx, $(_param_tuple.$i,)*).await?;
+                (self)(rqctx, $(_param_tuple.$i,)*).await.map_err(|e| HandlerError::Handler(Box::new(e)))?;
```

Comment:
It's necessary to erase the user's error type here, so that handler functions with arbitrary error return values can coexist in the same API.

We could alternatively do that by just going ahead and serializing the error response here. However, I didn't do that, because I thought it would be nicer to be able to give back a thing that also implements `fmt::Display`, for when we log that the request completes with an error. Thus, the `HandlerError` enum.